### DISCUSSION TITLE
Handle conflicted documents on start

### DISCUSSION THREAD

**Discussion by kaptnkoala** [cofu-app/cbl-dart#399] [Q&A]
We receive support mails from our users, that at some point the synchronization of the game state between their devices stop to work. 

We have now received a more concrete error description some days ago, where we could retrace analytics events to identify a situation where the user was not able to sync his game state, after an initial sync conflict was not fully resolved. 

In case of a sync conflict, we bring up a dialog, where the user can choose which of the conflicting game states should be selected. When the dialog gets popped (because of some other asynchronous navigation operation trying to recreate a certain route stack) or the app is closed or crashes, while the sync conflict is not resolved, the document seems to stay conflicted in the Couchbase Lite Layer. On the next start of the app, we see an info message that there is a conflicted document and also receive a 409 error, when trying to replicate this document. But on all subsequent app starts just it just shows the info message but not the error. 

```
I/flutter (17711): [couchbase] INFO: <replicator> {Repl#11} Scanning for pre-existing conflicts...
I/flutter (17711): [couchbase] INFO: <replicator> {Repl#11} Found 1 conflicted docs in 0.007 sec
I/flutter (17711): [couchbase] INFO: <replicator> {Repl#11} Read local checkpoint 'cp-T7d9tlvDn+jtBGat0EJS1P//Azc=': 
```

```
I/flutter (17711): [couchbase] ERROR: <replicator> {N8litecore4repl6PusherE#14} Got error response to rev 'game:B5FujJ3vbFYVTqiM2ja83Zwokx63' #123-8f0e20b714091b6bada5d5801ff31e28e7718dfe (seq #77): HTTP 409 'Document revision conflict' CBLError
I/flutter (17711): #0      CouchLogger.log (package:fourpics_shared/common/storage/couch_logger.dart:95:35)
I/flutter (17711): #1      _setupLogger.<anonymous closure> (package:cbl/src/log/logger.dart:203:12)
```

The game state synchronization for the user is broken from that point, until he dis- and reconnects his account, which essentially drops and recreates the conflicted document. Is there any chance, that we can access the the conflicted documents and resolve them later on? When looking at the API provided by CBL Dart and CBL I did not find a clear solution.

What we will definitely implement is a check for any unresolved conflict when the app switches to the background. In this case we will just do a best guess selection of the preferred game state. In this way we hopefully reduce the number of cases, where we will stay in an unresolved conflict state in the future. But this will not help the users, that currently have that state.

Maybe you have some ideas on that?

**Comment by blaugold** [comment]
I would expect that during the next replication, after the app terminates or crashes, conflict resolution is retried. 

This is similar to how conflict resolution should be retried during the next replication after a custom conflict resolver threw an exception. At least for exceptions, there is [documented behavior](https://docs.couchbase.com/couchbase-lite/current/c/conflict.html#important-guidelines-and-best-practices).

>If there is an exception thrown in the resolve() method, the exception will be caught and handled:
>
> * The conflict to resolve will be skipped. The pending conflicted documents will be resolved when the replicator is restarted.

I can't verify that any of one of those behaviors work, so I'll have to investigate.

Unfortunately, I don't think there currently is anything besides purging the document and pulling it from the server.

**Reply by blaugold** [reply]
Retrying conflict resolution currently does not work because of a bug in the Couchbase Lite C SDK (https://github.com/couchbase/couchbase-lite-C/issues/335).

**Reply by kaptnkoala** [reply]
Thanks a lot @blaugold for figuring this out!!

**Comment by kaptnkoala** [comment]
@blaugold The issue seems to be resolved on the Couchbase Lite repo now. Any chance that this finds its way into cbl-dart soon?

**Reply by blaugold** [reply]
It looks like the fix will be in CBL C [3.0.3](https://issues.couchbase.com/browse/CBL-3611). Given that 3.0.2 was released relatively recently, it might take a while until 3.0.3 is released.

**Reply by kaptnkoala** [reply]
Thanks for the info!
